core/services/gas/block_history_estimator.go

```diff
@@ -424,10 +536,34 @@ func (b *BlockHistoryEstimator) EffectiveGasPrice(block Block, tx Transaction) *
 			priorityFeePerGas = maxFeeMinusBaseFee
 		}
 
-		gasPrice := big.NewInt(0).Add(priorityFeePerGas, block.BaseFeePerGas)
-		return gasPrice
+		effectiveGasPrice := big.NewInt(0).Add(priorityFeePerGas, block.BaseFeePerGas)
+		return effectiveGasPrice
 	default:
-		b.logger.Warnw(fmt.Sprintf("BlockHistoryEstimator: unknown transaction type %v, falling back to gasPrice", tx.Type), "block", block, "tx", tx)
-		return tx.GasPrice
+		b.logger.Warnw(fmt.Sprintf("BlockHistoryEstimator: ignoring unknown transaction type %v", tx.Type), "block", block, "tx", tx)
+		return nil
+	}
+}
+
+func (b *BlockHistoryEstimator) EffectiveTipCap(block Block, tx Transaction) *big.Int {
+	switch tx.Type {
+	case 0x2:
+		return tx.MaxPriorityFeePerGas
+	default:
+		if !(tx.Type == 0x0 || tx.Type == 0x1) {
+			b.logger.Warnw(fmt.Sprintf("BlockHistoryEstimator: ignoring unknown transaction type %v", tx.Type), "block", block, "tx", tx)
+			return nil
+		}
+		if tx.GasPrice == nil {
+			return nil
+		}
+		if block.BaseFeePerGas == nil {
+			return nil
+		}
+		effectiveTipCap := big.NewInt(0).Sub(tx.GasPrice, block.BaseFeePerGas)
+		if effectiveTipCap.Cmp(big.NewInt(0)) < 0 {
+			b.logger.Warnw("BlockHistoryEstimator: invariant violated: GasPrice - BaseFeePerGas >= 0", "block", block, "tx", tx)
+			return nil
+		}
+		return effectiveTipCap
```

Comment:
What about splitting the default in to two cases?
```suggestion
	case 0x0, 0x1:
		if tx.GasPrice == nil {
			return nil
		}
		if block.BaseFeePerGas == nil {
			return nil
		}
		effectiveTipCap := big.NewInt(0).Sub(tx.GasPrice, block.BaseFeePerGas)
		if effectiveTipCap.Cmp(big.NewInt(0)) < 0 {
			b.logger.Warnw("BlockHistoryEstimator: invariant violated: GasPrice - BaseFeePerGas >= 0", "block", block, "tx", tx)
			return nil
		}
		return effectiveTipCap
	default:
		b.logger.Warnw(fmt.Sprintf("BlockHistoryEstimator: ignoring unknown transaction type %v", tx.Type), "block", block, "tx", tx)
		return nil
```